pints/tests/test_mcmc_haario_ac.py

```diff
@@ -0,0 +1,198 @@
+        cls.times = np.linspace(0, 1000, 1000)
+        cls.values = cls.model.simulate(cls.real_parameters, cls.times)
+
+        # Add noise
+        cls.noise = 10
+        cls.values += np.random.normal(0, cls.noise, cls.values.shape)
+        cls.real_parameters.append(cls.noise)
+        cls.real_parameters = np.array(cls.real_parameters)
+
+        # Create an object with links to the model and time series
+        cls.problem = pints.SingleOutputProblem(
+            cls.model, cls.times, cls.values)
+
+        # Create a uniform prior over both the parameters and the new noise
+        # variable
+        cls.log_prior = pints.UniformLogPrior(
+            [0.01, 400, cls.noise * 0.1],
+            [0.02, 600, cls.noise * 100]
+        )
+
+        # Create a log likelihood
+        cls.log_likelihood = pints.GaussianLogLikelihood(cls.problem)
+
+        # Create an un-normalised log-posterior (log-likelihood + log-prior)
+        cls.log_posterior = pints.LogPosterior(
+            cls.log_likelihood, cls.log_prior)
+
+    def test_method(self):
+
+        # Create mcmc
+        x0 = self.real_parameters * 1.1
+        mcmc = pints.HaarioACMCMC(x0)
+
+        # Configure
+        mcmc.set_target_acceptance_rate(0.3)
+        mcmc.set_initial_phase(True)
+
+        # Perform short run
+        rate = []
+        chain = []
+        for i in range(100):
+            x = mcmc.ask()
+            fx = self.log_posterior(x)
+            sample = mcmc.tell(fx)
+            if i == 20:
+                mcmc.set_initial_phase(False)
+            if i >= 50:
+                chain.append(sample)
+            rate.append(mcmc.acceptance_rate())
+            if np.all(sample == x):
+                self.assertEqual(mcmc.current_log_pdf(), fx)
+
+        chain = np.array(chain)
+        rate = np.array(rate)
+        self.assertEqual(chain.shape[0], 50)
+        self.assertEqual(chain.shape[1], len(x0))
+        self.assertEqual(rate.shape[0], 100)
+
+    def test_replace(self):
+
+        x0 = self.real_parameters * 1.1
+        mcmc = pints.HaarioACMCMC(x0)
+
+        # One round of ask-tell must have been run
+        self.assertRaisesRegex(
+            RuntimeError, 'already running', mcmc.replace, x0, 1)
+
+        mcmc.ask()
+
+        # One round of ask-tell must have been run
+        self.assertRaises(RuntimeError, mcmc.replace, x0, 1)
+
+        mcmc.tell(0.5)
+        mcmc.replace([1, 2, 3], 10)
+        mcmc.replace([1, 2, 3], 10)
+
+        # New position must have correct size
+        self.assertRaisesRegex(
+            ValueError, '`current` has the wrong dimensions',
+            mcmc.replace, [1, 2], 1)
+
+        # Proposal can be changed too
+        mcmc.ask()
+        mcmc.replace([1, 2, 3], 10, [3, 4, 5])
+
+        # New proposal must have correct size
+        self.assertRaisesRegex(
+            ValueError, '`proposed` has the wrong dimensions',
+            mcmc.replace, [1, 2, 3], 3, [3, 4])
+
+    def test_flow(self):
```

Comment:
Not sure how much (if any) of this is changed by Haario/Remi classes, and how much should go in single test for base class

Reply:
The problem is that the base class is just abstract. Would you call Haario or Remi in a test of the base class?

Reply:
Yup! It's not super nice, I agree, but I think it's the best compromise. We have similar things already in the tests for the OptimisationController and the tests for the MCMCController (both of which pick a single method and test with that)

Reply:
Yep, you're right. I have changed this now.